### PR TITLE
Dev (WIP)

### PR DESCRIPTION
**Pull request by erogol**
- [x] Tensorflow updates
   - [x] TF melgan implementation 
   - [x] Convert the latest model releases to TF
   - [x] Colab sample for TF inference
   - [x] TFLite conversion script for Tacotron2
   - [x] TFlite conversion script for Melgan
- [x] Adding a new VOCODER module. Now we can use Mozilla TTS directly to train the latest vocoder models and enable end-to-end faster than real-time inference performance.
    - [x] Melgan
    - [x] MultiBandMelgan
    - [x] Random Window Discriminator
    - [x] Enable native vocoder support in server.py
- [X] Masking intermediate model outputs against sequence padding. 
    - This aims to mitigate the distortion propagated by the convolution of padded parts of shorter sequences in a batch.
- [X] Double Decoder Consistency [Optional]
    - [Details](https://erogol.com/solving-attention-problems-of-tts-models-with-double-decoder-consistency/)
    - enable/disable using ```double_decoder_consistency``` field in `config.json`
    - set `ddc_r` in `config.json`. (try 7 as your initial run)
- [X] Refactoring Tacotron/Tacotron2 models based on a abstract class. 
- [X] Make masking of decoder/postnet outputs and attention weights optional. 
   - I observed that using masking is useful for a hard dataset but it hurts spectrogram predictions.  Therefore I made it optional to set it accordingly.
- [X] Big refactoring to a more consistent project tree. 
- [X] Add notebooks for the new models and samples.
- [X] OOM fix due to open figures in training.
- [X] Phoneme coverage notebook by @thorstenMueller
- [X] APEX float16 training by @repodiac 
- [X] Multi-speaker model by @Edresson

### CODE REVIEW

**Comment by Mic92** [comment]
@mweinelt and me are working on adding TTS to nixpkgs: https://github.com/mweinelt/nixpkgs/commits/tts 
That means all dependencies would be pre-compiled for NixOS, Linux distributions and macOS.

I have a few question regarding the dev version:

Which backend would be the recommend one? Pytorch or tensorflow? Also which one would be faster without CUDA?

**Comment by erogol** [comment]
We support PyTorch natively with all of our tools and try to provide conversion for torch models to TF. Therefore, it is easier to use PyTorch if you like to use this repo. Being said that, there are two options using Pytorch. You can export model using Torch script to C++ and do all the computation there which I'd say you might get faster results. TF also has different utilities for device-specific optimization, but I am not too familiar with them. If you decide to use TF, then you need to convert our vocoder model as it is done to Tacotron model and use them together in TF backend.

File: README.md [hunk 1]
```diff
@@ -84,7 +87,7 @@ Audio length is approximately 6 secs.
 
 
 ## Datasets and Data-Loading
-TTS provides a generic dataloder easy to use for new datasets. You need to write an preprocessor function to integrate your own dataset.Check ```datasets/preprocess.py``` to see some examples. After the function, you need to set ```dataset``` field in ```config.json```. Do not forget other data related fields too.  
+TTS provides a generic dataloder easy to use for new datasets. You need to write an preprocessor function to integrate your own dataset.Check ```datasets/preprocess.py``` to see some examples. After the function, you need to set ```dataset``` field in ```config.json```. Do not forget other data related fields too.
```
Comment (Mic92) [inline README.md:90]:
```suggestion
TTS provides a generic dataloader easy to use for new datasets. You need to write an preprocessor function to integrate your own dataset.Check ```datasets/preprocess.py``` to see some examples. After the function, you need to set ```dataset``` field in ```config.json```. Do not forget other data related fields too.
```

File: README.md [hunk 2]
```diff
@@ -3,9 +3,7 @@
 
 <img src="https://travis-ci.org/mozilla/TTS.svg?branch=dev"/>
 
-This project is a part of [Mozilla Common Voice](https://voice.mozilla.org/en). TTS aims a deep learning based Text2Speech engine, low in cost and high in quality. To begin with, you can hear a sample generated voice from [here](https://soundcloud.com/user-565970875/commonvoice-loc-sens-attn).
-
-TTS includes two different model implementations which are based on [Tacotron](https://arxiv.org/abs/1703.10135) and [Tacotron2](https://arxiv.org/abs/1712.05884). Tacotron is smaller, efficient and easier to train but Tacotron2 provides better results, especially when it is combined with a Neural vocoder. Therefore, choose depending on your project requirements.
+This project is a part of [Mozilla Common Voice](https://voice.mozilla.org/en). TTS aims a deep learning based Text2Speech engine, low in cost and high in quality. To begin with, you can hear a sample synthesized voice from [here](https://soundcloud.com/user-565970875/commonvoice-loc-sens-attn).
```
Comment (Mic92) [inline README.md:6]: Maybe its time to replace this with https://github.com/mozilla/TTS/blob/0176cd6b2c673ee5a9b2d7e3226dcf1cd37a82ee/tests/data/ljspeech/wavs/LJ001-0001.wav?raw=true 
since it sound so much better.

Comment (erogol) [reply README.md:6]: Good point. I'll update it with the new version.
